### PR TITLE
build: update bazel, upstream envoy, & ios rules

### PR DESCRIPTION
**Pull request by rebello95**
- Update bazel to 3.4.1 (depends on https://github.com/envoyproxy/envoy/pull/12123)
- Update envoy upstream to include ^
- Update rules apple and rules swift
- Update our header map iteration to fix breaking changes upstream from https://github.com/envoyproxy/envoy/pull/12103

Signed-off-by: Michael Rebello <me@michaelrebello.com>

### CODE REVIEW

File: bazel/swift_static_framework.bzl [hunk 1]
```diff
@@ -136,8 +136,8 @@ _swift_static_framework = rule(
             cfg = "host",
             executable = True,
         ),
-        _whitelist_function_transition = attr.label(
-            default = "@build_bazel_rules_apple//tools/whitelists/function_transition_whitelist",
+        _allowlist_function_transition = attr.label(
+            default = "@build_bazel_rules_apple//tools/allowlists/function_transition_allowlist",
```
Comment (buildbreaker) [inline bazel/swift_static_framework.bzl:140]: did bazel change these?

Comment (rebello95) [reply bazel/swift_static_framework.bzl:140]: Yea https://github.com/bazelbuild/rules_apple/commit/c283c82941e9f7e58f5176b68850e6a91b68e956

Comment (buildbreaker) [reply bazel/swift_static_framework.bzl:140]: 🙌

**Comment by rebello95** [comment]
Need Bazel 3.4.1: https://github.com/envoyproxy/envoy/pull/12123

File: bazel/swift_static_framework.bzl (hunk 1 above)
Comment (keith) [inline bazel/swift_static_framework.bzl:140]:
```suggestion
            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
```

**Comment by buildbreaker** [comment]
rerunning failed ci

**Comment by rebello95** [comment]
Looks like this is a legit failure when building the ObjC app: 

```
Execution platform: @local_config_platform//:host
examples/objective-c/hello_world/ViewController.m:1:9: fatal error: 'Envoy/Envoy-Swift.h' file not found
#import <Envoy/Envoy-Swift.h>
        ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target //examples/objective-c/hello_world:app failed to build
```

@keith do you happen to know of anything off the top of your head that may have broken our header setup with the rules updates?

**Comment by keith** [comment]
does the framework this produces have one at all after this change?

**Comment by rebello95** [comment]
Yep, it's present in `dist/Envoy.framework/Headers/Envoy-Swift.h`. Looks fine to me too 🤔

**Comment by keith** [comment]
how about the module map?

**Comment by rebello95** [comment]
Resolved with magic from @keith, more context: https://github.com/bazelbuild/bazel/issues/10674#issuecomment-658208918

**Pull request merged**